compiler/rustc_middle/src/middle/privacy.rs

```diff
@@ -104,30 +144,78 @@ impl<Id: Hash + Eq + Copy> AccessLevels<Id> {
     }
 
     pub fn set_access_level(&mut self, id: Id, tag: AccessLevel) {
-        let mut effective_vis = self.get_effective_vis(id).copied().unwrap_or_default();
+        let mut effective_vis = self
+            .get_effective_vis(id)
+            .copied()
+            .unwrap_or(EffectiveVisibility::from(Visibility::Restricted(id)));
         for level in [
             AccessLevel::Public,
             AccessLevel::Exported,
             AccessLevel::Reachable,
             AccessLevel::ReachableFromImplTrait,
         ] {
             if level <= tag {
-                *effective_vis.get_mut(level) = Some(Visibility::Public);
+                *effective_vis.get_mut(level) = Visibility::Public;
             }
         }
         self.map.insert(id, effective_vis);
     }
 
-    pub fn get_effective_vis(&self, id: Id) -> Option<&EffectiveVisibility> {
+    pub fn get_effective_vis(&self, id: Id) -> Option<&EffectiveVisibility<Id>> {
         self.map.get(&id)
     }
 
-    pub fn iter(&self) -> impl Iterator<Item = (&Id, &EffectiveVisibility)> {
+    pub fn iter(&self) -> impl Iterator<Item = (&Id, &EffectiveVisibility<Id>)> {
         self.map.iter()
     }
 
-    pub fn map_id<OutId: Hash + Eq + Copy>(&self, f: impl Fn(Id) -> OutId) -> AccessLevels<OutId> {
-        AccessLevels { map: self.map.iter().map(|(k, v)| (f(*k), *v)).collect() }
+    pub fn map_id<OutId: Hash + Eq + Copy>(
+        &self,
+        f: impl Fn(Id) -> OutId + Copy,
+    ) -> AccessLevels<OutId> {
+        AccessLevels { map: self.map.iter().map(|(k, v)| (f(*k), v.map_id(f))).collect() }
+    }
+
+    pub fn init_crate(&mut self, crate_id: Id) {
```

Comment:
I don't think this separate method is necessary, you could use `update` as with other nodes?